x-pack/plugin/search-business-rules/src/main/java/org/apache/lucene/search/CappedScoreQuery.java

```diff
@@ -0,0 +1,175 @@
+                            scorer.setMinCompetitiveScore(minScore);
+                        }
+                        
+                    });
+                }
+            };
+        }
+
+        @Override
+        public long cost() {
+            return bulkScorer.cost();
+        }
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        final Weight innerWeight = searcher.createWeight(query, ScoreMode.COMPLETE, 1f);
+        if (scoreMode.needsScores()) {
+            return new CappedScoreWeight(this, innerWeight, boost) {
+                @Override
+                public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
+                    final BulkScorer innerScorer = innerWeight.bulkScorer(context);
+                    if (innerScorer == null) {
+                        return null;
+                    }
+                    return new CappedBulkScorer(innerScorer, this, maxScore);
+                }
+
+                @Override
+                public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                    ScorerSupplier innerScorerSupplier = innerWeight.scorerSupplier(context);
+                    if (innerScorerSupplier == null) {
+                        return null;
+                    }
+                    return new ScorerSupplier() {
+                        @Override
+                        public Scorer get(long leadCost) throws IOException {
+                            final Scorer innerScorer = innerScorerSupplier.get(leadCost);
+                            // short-circuit if scores will not need capping
+                            innerScorer.advanceShallow(0);
+                            if (innerScorer.getMaxScore(DocIdSetIterator.NO_MORE_DOCS) <= maxScore) {
+                              return innerScorer;
+                            }                            
+                            return new CappedScorer(innerWeight, innerScorer, maxScore);
+                        }
+
+                        @Override
+                        public long cost() {
+                            return innerScorerSupplier.cost();
+                        }
+                    };
+                }
+
+                @Override
+                public Matches matches(LeafReaderContext context, int doc) throws IOException {
+                    return innerWeight.matches(context, doc);
+                }
+
+                @Override
+                public Scorer scorer(LeafReaderContext context) throws IOException {
+                    ScorerSupplier scorerSupplier = scorerSupplier(context);
+                    if (scorerSupplier == null) {
+                        return null;
+                    }
+                    return scorerSupplier.get(Long.MAX_VALUE);
+                }
```

Comment:
is there a reason why we can't have these implementations of scorer, scorerSupplier, etc. directly in CappedScoreWeight?